src/lex.cpp

```diff
@@ -1337,6 +1339,75 @@ void lexer::parse_number() {
     QLJS_SLOW_ASSERT(
         !(number_begin[0] == u8'0' && this->is_digit(number_begin[1])));
   }
+  if (!has_decimal_point && !has_exponent && !is_bigint) {
+    std::string cleaned_string = "";
+    for (char8 c : string8_view(number_begin, input - number_begin)) {
+      if (c != '_') {
+        cleaned_string.push_back(static_cast<char>(c));
+      }
+    }
+    if (cleaned_string.size() > 15) {
+      bool error = false;
+      bump_vector<char8, monotonic_allocator> num_pretty(
+          "opening_tag_name_pretty", &this->allocator_);
+      if (cleaned_string.size() > 309) {
+        num_pretty += u8"inf";
+        string8_view num_pretty_view(num_pretty);
+        this->diag_reporter_->report(diag_number_literal_will_lose_precision{
+            .characters = source_code_span(number_begin, input),
+            .rounded_val = num_pretty_view,
+        });
+        error = true;
+      }
+      if (!error) {
+        long double long_num = stold(cleaned_string);
+        double num = long_num;
+        std::array<char, 310> result_string;
+        int rc = std::snprintf(result_string.data(), result_string.size(),
+                               "%.0f", num);
```

Comment:
Why not `snprintf` the `long double` with `%Ld`?
```suggestion
        long double long_num = stold(cleaned_string);
        std::array<char, 310> result_string;
        int rc = std::snprintf(result_string.data(), result_string.size(),
                               "%.0Lf", long_num);
```

Reply:
The conversion from long double to double is actually intentional, as I need the lesser precision of a double. Originally, I used stod directly, but that would raise an out of range error for certain numbers with 309 digits or greater (numbers with 310 digits or greater can't reach this point in the code, though). I would catch this error, use stold, then assign the value of long_num to num to get that loss of precision without raising an error. Because I'm not allowed to use exception handling, I have to do this for every number now. However, a new problem is that clang doesn't like the implicit conversion from long double to double. What should I do about this? Personally, I think that we should add another check. If the number is greater than this magic constant:

179769313486231570814527423731704356798070567525844996598917476803157260780028538760589558632766878171540458953514382464234321326889464182768467546703537516986049910576551282076245490090389328944075868508455133942304583236903222948165808559332123348274797826204144723168738177180919299881250404026184124858368

then it will definitely lose precision. If we have this check, we can use stod directly and solve the style problem, the error handling problem, and use one less double. The downside is that this big 309 character string needs to be allocated. What are your thoughts?

Reply:
> However, a new problem is that clang doesn't like the implicit conversion from long double to double. What should I do about this?

Clang is fine with a conversion, as long as it's explicit. Use `static_cast`.

> The downside is that this big 309 character string needs to be allocated.

Why would we need to allocate that number as a string?

---

Keep in mind that `long double` differs between platforms. Sometimes it's 64 bits (same as `double`), sometimes it's 80 bits, and sometimes it's 128 bits. (Our CI only tests 64 and 80 bits.)